datashuttle/utils/directories.py

```diff
@@ -30,31 +29,174 @@ def make_dirs(paths: Union[Path, List[Path]]) -> None:
+                project, base_dir, "local", sub
+            )
+            existing_ses_values = utils.get_first_sub_ses_keys(
+                existing_ses_names
+            )
+
+            for new_ses in utils.get_first_sub_ses_keys(new_ses_names):
+
+                if new_ses in existing_ses_values:
+                    utils.raise_error(
+                        f"Cannot make directories. "
+                        f"The key ses-{new_ses} for {sub} already exists in the project"
+                    )
+
+
+# --------------------------------------------------------------------
+# Search Existing Dirs
+# --------------------------------------------------------------------
+
+# Search Subjects / Sessions
+# --------------------------------------------------------------------
+
+
+def search_sub_or_ses_level(
+    project: DataShuttle,
+    base_dir: Path,
+    local_or_remote: str,
+    sub: Optional[str] = None,
+    ses: Optional[str] = None,
+    search_str: str = "*",
 ) -> List[str]:
     """
-    Use glob to search the full search path (including prefix) with glob.
-    Files are filtered out of results, returning directories only.
+    Search project folder at the subject or session level
+
+    project: datashuttle project. Currently, this is used
+             as a holder for some ssh configs to avoid too many
+             arguments, but this is not nice and breaks the
+             general rule that these functions should operate
+             project-agnostic.
+
+    local_or_remote: search in local or remote project
+    sub: either a subject name (string) or None. If None, the search
+         is performed at the top_level_dir_name level
+    ses: either a session name (string) or None, This must not
+         be a session name if sub is None. If provided (with sub)
+         then the session dir is searched
+     str: glob-format search string to search at the
+          directory level.
     """
-    all_dirnames = []
-    for file_or_dir in glob.glob(search_path_with_prefix.as_posix()):
-        if os.path.isdir(file_or_dir):
-            all_dirnames.append(os.path.basename(file_or_dir))
-    return all_dirnames
+    if ses and not sub:
+        utils.raise_error(
+            "cannot pass session to "
+            "_search_sub_or_ses_level() without subject"
+        )
+
+    if sub:
+        base_dir = base_dir / sub
+
+    if ses:
+        base_dir = base_dir / ses
+
+    search_results = search_for_directories(
+        project, base_dir, local_or_remote, search_str
+    )
+    return search_results
+
+
+def search_data_dirs_sub_or_ses_level(
+    project, data_type_dirs, base_dir, local_or_remote, sub, ses=None
+):
+    """
+    Search  a subject or session directory specifically
+    for data_types. First searches for all folders / files
+    in the directory, and then returns any dirs that
+    match data_type name.
+
+    see _search_sub_or_ses_level() for inputs.
```

Comment:
You might want to restyle the docstrings and make more explicit the parameters names and their type, and the return type

Reply:
Thanks this has type hints in a [newer version](https://github.com/neuroinformatics-unit/datashuttle/blob/6e03f8be840459480aeabecfd849648cbc4f4065/datashuttle/utils/directories.py#L278) and I've made the docstring clearer